src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs

```diff
@@ -11,26 +13,64 @@ namespace Microsoft.AspNetCore.Http.Generators.StaticRouteHandlerModel;
 
 internal class Endpoint
 {
-    public string HttpMethod { get; }
-    public EndpointRoute Route { get; }
-    public EndpointResponse Response { get; }
-    public List<DiagnosticDescriptor> Diagnostics { get; } = new List<DiagnosticDescriptor>();
-
-    public (string, int) Location { get; }
-    public IInvocationOperation Operation { get; }
-
-    private WellKnownTypes WellKnownTypes { get; }
+    private string? _argumentListCache;
 
     public Endpoint(IInvocationOperation operation, WellKnownTypes wellKnownTypes)
     {
         Operation = operation;
-        WellKnownTypes = wellKnownTypes;
         Location = GetLocation();
         HttpMethod = GetHttpMethod();
-        Response = new EndpointResponse(Operation, wellKnownTypes);
-        Route = new EndpointRoute(Operation);
+
+        if (!operation.TryGetRouteHandlerPattern(out var routeToken))
+        {
+            Diagnostics.Add(DiagnosticDescriptors.UnableToResolveRoutePattern);
```

Comment:
I'm inclined to have each component of the model maintain its own diagnostics. As we build up the diagnostics we emit, it'll make it easier to warn for specific scenarios that might not be immediately accessible at the top-level.

For example, let's say that we find a `TryParse` method on a parameter that doesn't have the correct signature. It would be easier to read and track if we were to capture that diagnostic when we're constructing the parameter instead of bubbling it up and capturing them all at the top.

The downside is it does make the code to aggregate them all when we emit a little honky but that seems worth it for the benefits of having diagnostics localized to where they would be thrown (similar to an exception).

Reply:
I think it would be better to give EndpointParameter a backreference to Endpoints.Diagnostics if we want to capture the diagnostic while constructing the parameter. I just didn't like greedily creating a list for each small component even though it's not that expensive. The aggregation and repeated declarations were kind of verbose too as you say.